apollo-router/src/plugins/file_uploads/multipart_form_data.rs

```diff
@@ -0,0 +1,89 @@
+use bytes::Bytes;
+use bytes::BytesMut;
+use futures::stream::StreamExt;
+use futures::stream::TryStreamExt;
+use futures::Stream;
+use http::HeaderMap;
+use http::HeaderValue;
+use mediatype::names::BOUNDARY;
+use mediatype::names::FORM_DATA;
+use mediatype::names::MULTIPART;
+use mediatype::MediaType;
+use rand::RngCore;
+
+use super::MapField;
+use super::MultipartRequest;
+use super::UploadResult;
+
+pub(super) struct MultipartFormData {
+    boundary: String,
+    operations: hyper::Body,
+    map: MapField,
+    multipart: MultipartRequest,
+}
+
+impl MultipartFormData {
+    pub(super) fn new(operations: hyper::Body, map: MapField, multipart: MultipartRequest) -> Self {
+        let boundary = format!(
+            "------------------------{:016x}",
+            rand::thread_rng().next_u64()
+        );
+        Self {
+            boundary,
+            operations,
+            map,
+            multipart,
+        }
+    }
+
+    pub(super) fn content_type(&self) -> HeaderValue {
+        let boundary =
+            mediatype::Value::new(&self.boundary).expect("boundary should be valid value");
+        let params = [(BOUNDARY, boundary)];
+        let mime = MediaType::from_parts(MULTIPART, FORM_DATA, None, &params);
+        mime.to_string()
+            .try_into()
+            .expect("mime should be valid header value")
+    }
+
+    pub(super) async fn into_stream(mut self) -> impl Stream<Item = UploadResult<Bytes>> {
+        let map_bytes = serde_json::to_vec(&self.map).expect("map should be serializable to JSON");
+        let field_prefix = |name: &str| {
+            tokio_stream::once(Ok(Bytes::from(format!(
+                "--{}\r\nContent-Disposition: form-data; name=\"{}\"\r\n\r\n",
+                self.boundary, name
+            ))))
+        };
+
+        let static_part = field_prefix("operations")
+            .chain(self.operations.map_err(Into::into))
+            .chain(tokio_stream::once(Ok("\r\n".into())))
+            .chain(field_prefix("map"))
+            .chain(tokio_stream::once(Ok(Bytes::from(map_bytes))))
+            .chain(tokio_stream::once(Ok("\r\n".into())));
+        let last = tokio_stream::once(Ok(format!("\r\n--{}--\r\n", self.boundary).into()));
+
+        let file_names = self.map.into_keys().collect();
+        let boundary = self.boundary;
+        let file_prefix = move |headers: &HeaderMap| {
+            let mut prefix = BytesMut::new();
+            prefix.extend_from_slice(b"\r\n--");
+            prefix.extend_from_slice(boundary.as_bytes());
+            prefix.extend_from_slice(b"\r\n");
+            for (k, v) in headers.iter() {
+                prefix.extend_from_slice(k.as_str().as_bytes());
+                prefix.extend_from_slice(b": ");
+                prefix.extend_from_slice(v.as_bytes());
+                prefix.extend_from_slice(b"\r\n");
+            }
```

Comment:
this way to write it will result in each `chain` call sending a separate chunk, even the small ones. We should try to send larger chunks at once, to be mindful of the subgraph side, that will need to reparse the accumulated chunks until they get to the boundary. If the entire file is small enough to send as one chunk with the boundaries, great, if it is larger it can be cut up in pieces

Reply:
fixed in [367ea94](https://github.com/apollographql/router/pull/4584/commits/367ea94546ff3f4db9c41d7e0f81e9884beddc10)